### ISSUE TITLE
Transition to CLOSED state does not complete recovery future

### ISSUE THREAD

**Issue by vyzo** [atomix/copycat#237]
We are running into a problem when running under heavy load. Tracing the logs indicate that the following events happen:
- CopycatClient emits onStateChange with new state SUSPENDED, triggering a recovery in our code
- CopycatClient then emits onStateChange with new state DISCONNECTED
- the recovery initiated by the first event does not complete; the thread waiting on the completion of the recovery hangs

Unfortunately I can't make a small reproducible case, due to the intermittent nature of the problem - it can take an hour for it to occur, which suggests a race condition.

**Comment by vyzo** [comment]
Note that we are using copycat-1.1.4, which is the current release.

**Comment by vyzo** [comment]
An update about further problems; We implemented a simple workaround which would cancel pending recoveries when this event happened, and proceed to reconnect the client.

This works for a while, but eventually a new terminal state appears on the client: It successfully returns from the reconnect, but never transitions to CONNECTED state.

**Comment by madjam** [comment]
> We implemented a simple workaround which would cancel pending recoveries when this event happened, and proceed to reconnect the client.

By default copycat has automatic client (session) recovery enabled. Any particular reason why you are doing manual client recovery?

**Comment by vyzo** [comment]
The automatic recovery never gives up on failures, and we do want to eventually fail if the cluster is unreachable.
We are using a custom recovery strategy, with a randomized exponential attempt that gives up after 10 tries. At a second level we also do client reconnects (for another 5 tries, randomized exponential backoff).

Note that the second issue we are seeing with the cancellation workaround is a reconnect succeeding, but without generating a state change event, so it might be an instance of the same bug (I can certainly make another issue if you like). This happens intermittently under high write load.

**Comment by vyzo** [comment]
Actually, the historical reason why we are not using the automated recovery is that it would generate an IllegalStateException and fail to properly shutdown when the client was disconnected by calling close.

**Comment by madjam** [comment]
`IllegalStateException` on calling `close` feels like a bug to me. It also looks like you are running with the latest copycat release. So this definitely is something we need to address.

Do you happen to have a stack trace for that error?

**Comment by vyzo** [comment]
It's been a while since we resolved to use custom recovery for clean shutdowns, so I don't have any stack traces handy. What I recall was happening is that the recovery strategy would be invoked on explicit close, and try to recover the closed session resulting in the exception.
